lib/dash/content_protection.js

```diff
@@ -250,6 +250,179 @@ shaka.dash.ContentProtection.parseFromRepresentation = function(
   return repContext.defaultKeyId || context.defaultKeyId;
 };
 
+/**
+ * Gets a Widevine license URL from a content protection element
+ * containing a custom `ms:laurl` element
+ *
+ * @param {shaka.dash.ContentProtection.Element} element
+ * @return {string}
+ * @private
+ */
+shaka.dash.ContentProtection.getWidevineLicenseUrl_ = function(element) {
+  var mslaurlNode = shaka.util.XmlUtils.findChild(element.node, 'ms:laurl');
+
+  if (mslaurlNode) {
+    return mslaurlNode.getAttribute('licenseUrl') || '';
+  }
+
+  return '';
+};
+
+/**
+ * @typedef {{
+ *   type: number,
+ *   length: number,
+ *   value: string
+ * }}
+ *
+ * @description
+ * The parsed result of a PlayReady object record.
+ *
+ * @property {number} type
+ *   Type of data stored in the record.
+ * @property {number} length
+ *   Size of the record in bytes.
+ * @property {string} value
+ *   Record content.
+ */
+shaka.dash.ContentProtection.PlayReadyRecord;
+
+/**
+ * @typedef {{
+ *   length: number,
+ *   recordCount: number,
+ *   records: Array.<shaka.dash.ContentProtection.PlayReadyRecord>,
+ * }}
+ *
+ * @description
+ * The parsed result of a PlayReady object.
+ *
+ * @property {number} length
+ *   Size of the PlayReady header object in bytes.
+ * @property {number} recordCount
+ *   Number of records in the PlayReady object.
+ * @property {Array.<shaka.dash.ContentProtection.PlayReadyRecord>} records
+ *   Contains a variable number of records that contain license acquisition information.
+ */
+shaka.dash.ContentProtection.PlayReadyHeaderObject;
+
+/**
+ * A map of PlayReady record types to description.
+ *
+ * @const {!Object.<number, string>}
+ * @private
+ */
+var PLAYREADY_RECORD_TYPES = {
+  0x0001 : 'RIGHTS_MANAGEMENT',
+  0x0002 : 'RESERVED',
+  0x0003 : 'EMBEDDED_LICENSE'
+};
+
+/**
+ * @param {Uint16Array} recordData
+ * @param {number} recordCount
+ * @return {Array.<shaka.dash.ContentProtection.PlayReadyRecord>}
+ * @private
+ */
+shaka.dash.ContentProtection.parseRecords_ = function(recordData, recordCount) {
+  var head = 0;
+  var records = [];
+  for (var i = 0; i < recordCount; i++) {
+    var recordRaw = recordData.subarray(head);
+
+    var type = recordRaw[0];
+    var length = recordRaw[1];
```

Comment:
Here too, there are unchecked indexes. Can we add some checks to insure that the right amount of data is present before indexing?